# 01 - Data Engineering: Ingesta y Limpieza

**Objetivo:** Descargar, limpiar y preparar los datos macroeconómicos para análisis.

**Fuentes:**
- FRED API: 11 series macroeconómicas
- World Bank: Gold Price (Pink Sheet)

**Output:** `data/processed/dataset_modelo.csv` (12 variables, período 1992+)

## 1. Setup

In [1]:
# Librerías
import pandas as pd
import numpy as np
import requests
import openpyxl
import os
from dotenv import load_dotenv

# Directorio de trabajo
os.chdir("/Users/javiermondragon/Documents/data_projects/inflation-predictor")

# Cargar API key
load_dotenv()
api_key = os.getenv("FRED_API_KEY")
print(f"API Key cargada: {api_key[:5]}...")

# Crear carpetas si no existen
os.makedirs("data/raw", exist_ok=True)
os.makedirs("data/processed", exist_ok=True)
print("Setup completo")

API Key cargada: b1ab9...
Setup completo


## 2. Definición de Variables

| Variable | Código FRED | Descripción | Frecuencia |
|----------|-------------|-------------|------------|
| CPI | CPIAUCSL | Índice de Precios al Consumidor (objetivo) | Mensual |
| Fed Rate | FEDFUNDS | Tasa de Fondos Federales | Mensual |
| Oil Price | DCOILWTICO | Precio del Petróleo WTI | Diario→Mensual |
| Unemployment | UNRATE | Tasa de Desempleo | Mensual |
| Industrial Prod. | INDPRO | Índice de Producción Industrial | Mensual |
| Money Supply M2 | M2SL | Oferta Monetaria M2 | Mensual |
| Retail Sales | RRSFS | Ventas Minoristas | Mensual |
| Capacity Util. | TCU | Utilización de Capacidad Instalada | Mensual |
| Treasury 10Y | GS10 | Rendimiento del Tesoro a 10 años | Mensual |
| PPI | PPIACO | Índice de Precios al Productor | Mensual |
| Consumer Sentiment | UMCSENT | Índice de Sentimiento del Consumidor | Mensual |
| Gold Price | World Bank | Precio del Oro (Pink Sheet) | Mensual |

## 3. Descarga de Datos

### 3.1 Series de FRED

In [2]:
# Definir series de FRED (11 variables)
series_fred = {
    "CPIAUCSL": "cpi",
    "FEDFUNDS": "fed_rate",
    "DCOILWTICO": "oil_price",
    "UNRATE": "unemployment",
    "INDPRO": "industrial_production",
    "M2SL": "money_supply_m2",
    "RRSFS": "retail_sales",
    "TCU": "capacity_utilization",
    "GS10": "treasury_10y",
    "PPIACO": "ppi",
    "UMCSENT": "consumer_sentiment"
}

url = "https://api.stlouisfed.org/fred/series/observations"

# Verificar cada serie
print("Verificando series de FRED...\n")
print(f"{'Código':<15} {'Nombre':<25} {'Registros'}")
print("="*50)

for codigo, nombre in series_fred.items():
    params = {
        "series_id": codigo,
        "api_key": api_key,
        "file_type": "json"
    }
    response = requests.get(url, params=params)
    data = response.json()
    
    if "observations" in data:
        print(f"✓ {codigo:<13} {nombre:<25} {len(data['observations'])}")
    else:
        print(f"✗ {codigo:<13} ERROR: {data.get('error_message')}")

Verificando series de FRED...

Código          Nombre                    Registros


✓ CPIAUCSL      cpi                       951


✓ FEDFUNDS      fed_rate                  861


✓ DCOILWTICO    oil_price                 10508
✓ UNRATE        unemployment              939


✓ INDPRO        industrial_production     1287


✓ M2SL          money_supply_m2           806


✓ RRSFS         retail_sales              410
✓ TCU           capacity_utilization      711


✓ GS10          treasury_10y              876


✓ PPIACO        ppi                       1359
✓ UMCSENT       consumer_sentiment        880


In [3]:
# Descargar y guardar datos crudos
print("Descargando datos de FRED...\n")

for codigo, nombre in series_fred.items():
    params = {
        "series_id": codigo,
        "api_key": api_key,
        "file_type": "json"
    }
    
    response = requests.get(url, params=params)
    data = response.json()
    
    df = pd.DataFrame(data["observations"])
    df.to_csv(f"data/raw/{nombre}_raw.csv", index=False)
    
    print(f"✓ {nombre}_raw.csv guardado ({len(df)} filas)")

print("\n" + "="*50)
print("Descarga de FRED completada: 11 series")

Descargando datos de FRED...



✓ cpi_raw.csv guardado (951 filas)


✓ fed_rate_raw.csv guardado (861 filas)


✓ oil_price_raw.csv guardado (10508 filas)


✓ unemployment_raw.csv guardado (939 filas)


✓ industrial_production_raw.csv guardado (1287 filas)


✓ money_supply_m2_raw.csv guardado (806 filas)
✓ retail_sales_raw.csv guardado (410 filas)


✓ capacity_utilization_raw.csv guardado (711 filas)


✓ treasury_10y_raw.csv guardado (876 filas)


✓ ppi_raw.csv guardado (1359 filas)


✓ consumer_sentiment_raw.csv guardado (880 filas)

Descarga de FRED completada: 11 series


### 3.2 Oro del World Bank

In [4]:
# Cargar oro del World Bank Pink Sheet
# NOTA: Descargar CMO-Historical-Data-Monthly.xlsx de World Bank y copiar a data/raw/

df_gold_wb = pd.read_excel('data/raw/CMO-Historical-Data-Monthly.xlsx', 
                           sheet_name='Monthly Prices', 
                           header=4)

# Procesar
df_gold_wb = df_gold_wb.rename(columns={'Unnamed: 0': 'date'})
df_gold_wb = df_gold_wb[['date', 'Gold']].copy()
df_gold_wb = df_gold_wb.iloc[1:]  # Quitar fila de unidades

# Convertir tipos
df_gold_wb['date'] = pd.to_datetime(df_gold_wb['date'].str.replace('M', '-'), format='%Y-%m')
df_gold_wb['Gold'] = pd.to_numeric(df_gold_wb['Gold'], errors='coerce')
df_gold_wb = df_gold_wb.rename(columns={'Gold': 'value'})

# Guardar
df_gold_wb.to_csv('data/raw/gold_price_raw.csv', index=False)

print(f"✓ gold_price_raw.csv guardado ({len(df_gold_wb)} filas)")
print(f"  Rango: {df_gold_wb['date'].min().strftime('%Y-%m')} a {df_gold_wb['date'].max().strftime('%Y-%m')}")

✓ gold_price_raw.csv guardado (792 filas)
  Rango: 1960-01 a 2025-12


## 4. Perfilamiento de Datos Crudos

In [5]:
# Cargar todos los datasets crudos
datasets = {}
nombres_archivos = list(series_fred.values()) + ['gold_price']

for nombre in nombres_archivos:
    datasets[nombre] = pd.read_csv(f'data/raw/{nombre}_raw.csv')

# Perfil de cada dataset
print("PERFIL DE DATOS CRUDOS")
print("="*70)
print(f"\n{'Variable':<25} {'Filas':<10} {'Fecha inicio':<15} {'Fecha fin'}")
print("-"*70)

for nombre, df in datasets.items():
    # Convertir fecha
    df['date'] = pd.to_datetime(df['date'])
    fecha_min = df['date'].min().strftime('%Y-%m')
    fecha_max = df['date'].max().strftime('%Y-%m')
    print(f"{nombre:<25} {len(df):<10} {fecha_min:<15} {fecha_max}")

PERFIL DE DATOS CRUDOS

Variable                  Filas      Fecha inicio    Fecha fin
----------------------------------------------------------------------
cpi                       951        1947-01         2026-03
fed_rate                  861        1954-07         2026-03
oil_price                 10508      1986-01         2026-04
unemployment              939        1948-01         2026-03
industrial_production     1287       1919-01         2026-03
money_supply_m2           806        1959-01         2026-02
retail_sales              410        1992-01         2026-02
capacity_utilization      711        1967-01         2026-03
treasury_10y              876        1953-04         2026-03
ppi                       1359       1913-01         2026-03
consumer_sentiment        880        1952-11         2026-02
gold_price                792        1960-01         2025-12


## 5. Identificación de Problemas

In [6]:
print("IDENTIFICACIÓN DE PROBLEMAS")
print("="*70)

for nombre, df in datasets.items():
    print(f"\n{nombre.upper()}")
    
    # Determinar columna de valor (FRED usa 'value', gold_price ya tiene 'value' después de procesar)
    if 'value' not in df.columns:
        print(f"  ⚠ Columna 'value' no encontrada. Columnas: {list(df.columns)}")
        continue
    
    # Verificar valores no numéricos en 'value'
    df['value'] = df['value'].astype(str)
    non_numeric = df[~df['value'].str.match(r'^-?\d+\.?\d*$')]['value'].unique()
    
    if len(non_numeric) > 0:
        print(f"  ⚠ Valores no numéricos: {list(non_numeric)}")
        print(f"    Cantidad: {len(df[df['value'].isin(non_numeric)])}")
    else:
        print(f"  ✓ Sin valores no numéricos")
    
    # Verificar nulos
    nulos = df['value'].isna().sum()
    if nulos > 0:
        print(f"  ⚠ Valores nulos: {nulos}")
    else:
        print(f"  ✓ Sin valores nulos")

IDENTIFICACIÓN DE PROBLEMAS

CPI
  ⚠ Valores no numéricos: ['.']
    Cantidad: 1
  ✓ Sin valores nulos

FED_RATE
  ✓ Sin valores no numéricos
  ✓ Sin valores nulos

OIL_PRICE
  ⚠ Valores no numéricos: ['.']
    Cantidad: 370
  ✓ Sin valores nulos

UNEMPLOYMENT
  ⚠ Valores no numéricos: ['.']
    Cantidad: 1
  ✓ Sin valores nulos

INDUSTRIAL_PRODUCTION
  ✓ Sin valores no numéricos
  ✓ Sin valores nulos

MONEY_SUPPLY_M2
  ✓ Sin valores no numéricos
  ✓ Sin valores nulos

RETAIL_SALES
  ⚠ Valores no numéricos: ['.']
    Cantidad: 1
  ✓ Sin valores nulos

CAPACITY_UTILIZATION
  ✓ Sin valores no numéricos
  ✓ Sin valores nulos

TREASURY_10Y
  ✓ Sin valores no numéricos
  ✓ Sin valores nulos

PPI
  ✓ Sin valores no numéricos
  ✓ Sin valores nulos

CONSUMER_SENTIMENT
  ⚠ Valores no numéricos: ['.']
    Cantidad: 210
  ✓ Sin valores nulos

GOLD_PRICE
  ✓ Sin valores no numéricos
  ✓ Sin valores nulos


## 6. Limpieza de Datos

In [7]:
def limpiar_serie_fred(nombre):
    """Limpia una serie de FRED: convierte tipos y maneja valores faltantes."""
    df = pd.read_csv(f'data/raw/{nombre}_raw.csv')
    
    # Seleccionar columnas relevantes
    df = df[['date', 'value']].copy()
    
    # Convertir fecha
    df['date'] = pd.to_datetime(df['date'])
    
    # Reemplazar '.' por NaN y convertir a numérico
    df['value'] = df['value'].replace('.', np.nan)
    df['value'] = pd.to_numeric(df['value'], errors='coerce')
    
    # Renombrar columna value
    df = df.rename(columns={'value': nombre})
    
    return df

# Limpiar todas las series FRED
print("Limpiando series de FRED...\n")

series_limpias = {}
for codigo, nombre in series_fred.items():
    series_limpias[nombre] = limpiar_serie_fred(nombre)
    n_validos = series_limpias[nombre][nombre].notna().sum()
    print(f"✓ {nombre}: {n_validos} valores válidos")

print("\nLimpieza de FRED completada")

Limpiando series de FRED...

✓ cpi: 950 valores válidos
✓ fed_rate: 861 valores válidos
✓ oil_price: 10138 valores válidos
✓ unemployment: 938 valores válidos
✓ industrial_production: 1287 valores válidos
✓ money_supply_m2: 806 valores válidos
✓ retail_sales: 409 valores válidos
✓ capacity_utilization: 711 valores válidos
✓ treasury_10y: 876 valores válidos
✓ ppi: 1359 valores válidos
✓ consumer_sentiment: 670 valores válidos

Limpieza de FRED completada


In [8]:
# Limpiar oro (World Bank)
df_gold = pd.read_csv('data/raw/gold_price_raw.csv')
df_gold['date'] = pd.to_datetime(df_gold['date'])
df_gold = df_gold.rename(columns={'value': 'gold_price'})

print(f"✓ gold_price: {df_gold['gold_price'].notna().sum()} valores válidos")

✓ gold_price: 792 valores válidos


## 7. Agregación a Frecuencia Mensual

La serie de petróleo (DCOILWTICO) tiene frecuencia diaria. Se agrega a mensual usando el **promedio mensual**.

In [9]:
# Agregar oil_price a frecuencia mensual
df_oil = series_limpias['oil_price'].copy()

# Crear columna de mes
df_oil['month'] = df_oil['date'].dt.to_period('M').dt.to_timestamp()

# Agregar por promedio mensual
df_oil_mensual = df_oil.groupby('month').agg({'oil_price': 'mean'}).reset_index()
df_oil_mensual = df_oil_mensual.rename(columns={'month': 'date'})

print(f"Oil price agregado: {len(df_oil)} días → {len(df_oil_mensual)} meses")

# Actualizar en series_limpias
series_limpias['oil_price'] = df_oil_mensual

Oil price agregado: 10508 días → 484 meses


## 8. Integración de Series

In [10]:
# Unir todas las series
print("Integrando series...\n")

# Empezar con CPI
df_final = series_limpias['cpi'].copy()

# Unir resto de series FRED
for nombre in series_fred.values():
    if nombre != 'cpi':
        df_final = df_final.merge(series_limpias[nombre], on='date', how='outer')

# Unir oro
df_final = df_final.merge(df_gold, on='date', how='outer')

# Ordenar por fecha
df_final = df_final.sort_values('date').reset_index(drop=True)

print(f"Dataset integrado: {len(df_final)} filas, {len(df_final.columns)} columnas")
print(f"Rango completo: {df_final['date'].min().strftime('%Y-%m')} a {df_final['date'].max().strftime('%Y-%m')}")
print(f"\nColumnas: {list(df_final.columns)}")

Integrando series...

Dataset integrado: 1360 filas, 13 columnas
Rango completo: 1913-01 a 2026-04

Columnas: ['date', 'cpi', 'fed_rate', 'oil_price', 'unemployment', 'industrial_production', 'money_supply_m2', 'retail_sales', 'capacity_utilization', 'treasury_10y', 'ppi', 'consumer_sentiment', 'gold_price']


In [11]:
# Ver nulos por columna
print("VALORES NULOS POR COLUMNA")
print("="*50)
print(df_final.isna().sum())
print(f"\nTotal filas: {len(df_final)}")

VALORES NULOS POR COLUMNA
date                       0
cpi                      410
fed_rate                 499
oil_price                876
unemployment             422
industrial_production     73
money_supply_m2          554
retail_sales             951
capacity_utilization     649
treasury_10y             484
ppi                        1
consumer_sentiment       690
gold_price               568
dtype: int64

Total filas: 1360


In [12]:
# Restringir al período con cobertura completa de todas las series.
#
# - Límite inferior: retail_sales (RRSFS) es la serie más corta, inicia 1992-01.
# - Límite superior: gold_price (World Bank Pink Sheet) es la serie con último
#   dato publicado más temprano. Forward-fill solo se aplica para tapar
#   huecos *internos*; no se extrapola más allá del último mes con dato real
#   en gold para evitar propagar 3+ meses de ffill sin justificación.
fecha_min = pd.Timestamp('1992-01-01')
fecha_max = df_gold['date'].max()

df_modelo = df_final[(df_final['date'] >= fecha_min) &
                     (df_final['date'] <= fecha_max)].copy()
df_modelo = df_modelo.sort_values('date').reset_index(drop=True)

print(f"Ventana seleccionada: {fecha_min:%Y-%m} → {fecha_max:%Y-%m}")
print(f"Filas en la ventana: {len(df_modelo)}")

# Diagnóstico de nulos internos antes de imputación
nulos_por_columna = df_modelo.isna().sum()
n_nulos_antes = nulos_por_columna.sum()
print(f"\nNulos internos antes de imputación: {n_nulos_antes}")
if n_nulos_antes > 0:
    print("Columnas con nulos:")
    print(nulos_por_columna[nulos_por_columna > 0])

# Imputación hacia adelante (forward-fill) para huecos internos.
#
# Motivación: el CPI de octubre 2025 no fue publicado por el BLS al momento
# de descarga (probable retraso asociado al shutdown del gobierno de EE.UU.
# de fines de 2025). Forward-fill preserva la secuencialidad temporal y
# evita usar información futura (ver propuesta, nota al pie 5), por lo que
# es coherente con la metodología declarada.
df_modelo = df_modelo.ffill()

# Eliminar filas residuales con nulos (si alguna serie empezara después de
# 1992-01, sus primeras observaciones no pueden imputarse con forward-fill
# y deben descartarse).
df_modelo = df_modelo.dropna().reset_index(drop=True)

n_nulos_despues = df_modelo.isna().sum().sum()
print(f"\nNulos después de forward-fill: {n_nulos_despues}")
print(f"Observaciones finales: {len(df_modelo)}")
print(f"Período efectivo: {df_modelo['date'].min():%Y-%m} a {df_modelo['date'].max():%Y-%m}")

Ventana seleccionada: 1992-01 → 2025-12
Filas en la ventana: 408

Nulos internos antes de imputación: 3
Columnas con nulos:
cpi             1
unemployment    1
retail_sales    1
dtype: int64

Nulos después de forward-fill: 0
Observaciones finales: 408
Período efectivo: 1992-01 a 2025-12


## 9. Validación del Dataset

In [13]:
# Estadísticas descriptivas del dataset final
print("ESTADÍSTICAS DESCRIPTIVAS DEL DATASET FINAL")
print("="*70)
print(df_modelo.describe().round(2))

ESTADÍSTICAS DESCRIPTIVAS DEL DATASET FINAL
                                date     cpi  fed_rate  oil_price  \
count                            408  408.00    408.00     408.00   
mean   2008-12-15 17:17:38.823529472  214.78      2.64      53.17   
min              1992-01-01 00:00:00  138.30      0.05      11.35   
25%              2000-06-23 12:00:00  172.58      0.19      25.45   
50%              2008-12-16 12:00:00  214.76      2.30      51.77   
75%              2017-06-08 12:00:00  244.21      4.94      75.43   
max              2025-12-01 00:00:00  326.03      6.54     133.88   
std                              NaN   49.67      2.17      28.94   

       unemployment  industrial_production  money_supply_m2  retail_sales  \
count        408.00                 408.00           408.00        408.00   
mean           5.64                  92.00          9902.05     170988.85   
min            3.40                  61.46          3381.20     114052.00   
25%            4.30       

In [14]:
# Verificar continuidad temporal (sin gaps)
# Nota: el CPI de octubre 2025 fue imputado con forward-fill en la etapa
# anterior, por lo que no debe aparecer aquí ningún gap > 32 días.
df_modelo['date_diff'] = df_modelo['date'].diff()
gaps = df_modelo[df_modelo['date_diff'] > pd.Timedelta(days=32)]

if len(gaps) > 0:
    print("⚠ GAPS TEMPORALES DETECTADOS:")
    print(gaps[['date', 'date_diff']])
else:
    print("✓ Serie temporal continua (sin gaps)")

# Eliminar columna auxiliar
df_modelo = df_modelo.drop(columns=['date_diff'])

✓ Serie temporal continua (sin gaps)


In [15]:
# Vista previa
print("PRIMERAS FILAS:")
df_modelo.head()

PRIMERAS FILAS:


,date,cpi,fed_rate,oil_price,unemployment,industrial_production,money_supply_m2,retail_sales,capacity_utilization,treasury_10y,ppi,consumer_sentiment,gold_price
0,1992-01-01,138.3,4.03,18.785455,7.3,61.4616,3381.2,115095.0,79.3278,7.03,115.6,67.5,354.45
1,1992-02-01,138.6,4.06,19.012500,7.4,61.8955,3400.0,114855.0,79.7329,7.34,116.0,68.8,353.91
2,1992-03-01,139.1,3.98,18.921818,7.4,62.4232,3403.9,114052.0,80.2464,7.54,116.1,76.0,344.34
3,1992-04-01,139.4,3.73,20.230000,7.4,62.8970,3399.7,114721.0,80.6802,7.48,116.3,77.2,338.62
4,1992-05-01,139.7,3.82,20.975500,7.6,63.1040,3398.6,114868.0,80.7640,7.39,117.2,79.2,337.24


In [16]:
# Últimas filas
print("ÚLTIMAS FILAS:")
df_modelo.tail()

ÚLTIMAS FILAS:


,date,cpi,fed_rate,oil_price,unemployment,industrial_production,money_supply_m2,retail_sales,capacity_utilization,treasury_10y,ppi,consumer_sentiment,gold_price
403,2025-08-01,323.291,4.33,64.864286,4.3,101.6247,22086.8,226329.0,76.1055,4.26,262.110,58.2,3368.03
404,2025-09-01,324.245,4.22,63.959048,4.4,101.6680,22169.8,225814.0,76.0461,4.12,262.054,55.1,3667.68
405,2025-10-01,324.245,4.09,60.894545,4.4,101.2195,22250.4,225814.0,75.6194,4.06,260.591,53.6,4058.33
406,2025-11-01,325.063,3.88,60.062222,4.5,101.0810,22296.5,226023.0,75.4249,4.09,261.914,51.0,4087.19
407,2025-12-01,326.031,3.72,57.972273,4.4,101.6113,22386.9,225352.0,75.7296,4.14,261.349,52.9,4309.23


## 10. Guardar Dataset Procesado

In [17]:
# Guardar dataset final
df_modelo.to_csv('data/processed/dataset_modelo.csv', index=False)

print("="*70)
print("DATASET GUARDADO EXITOSAMENTE")
print("="*70)
print(f"\nArchivo: data/processed/dataset_modelo.csv")
print(f"\nResumen:")
print(f"  • Filas: {len(df_modelo)}")
print(f"  • Columnas: {len(df_modelo.columns)}")
print(f"  • Período: {df_modelo['date'].min().strftime('%Y-%m')} a {df_modelo['date'].max().strftime('%Y-%m')}")
print(f"\nVariables:")
for i, col in enumerate(df_modelo.columns, 1):
    print(f"  {i:2}. {col}")

DATASET GUARDADO EXITOSAMENTE

Archivo: data/processed/dataset_modelo.csv

Resumen:
  • Filas: 408
  • Columnas: 13
  • Período: 1992-01 a 2025-12

Variables:
   1. date
   2. cpi
   3. fed_rate
   4. oil_price
   5. unemployment
   6. industrial_production
   7. money_supply_m2
   8. retail_sales
   9. capacity_utilization
  10. treasury_10y
  11. ppi
  12. consumer_sentiment
  13. gold_price


## Resumen

**Fuentes procesadas:**
- 11 series de FRED API
- 1 serie de World Bank (Gold, Pink Sheet)

**Transformaciones aplicadas:**
- Conversión de tipos de datos
- Manejo de valores no numéricos ('.' → NaN)
- Agregación de oil_price diario a mensual (promedio)
- Integración mediante merge por fecha
- Filtrado a período con cobertura completa (desde 1992-01, donde inicia retail_sales)
- **Imputación forward-fill para huecos internos** (caso octubre 2025: CPI no publicado por el BLS al momento de descarga; se imputa con el último valor válido disponible, consistente con la metodología declarada en la propuesta)

**Dataset de salida:** `data/processed/dataset_modelo.csv` — 12 variables, 408 observaciones mensuales (ene-1992 → dic-2025), sin nulos.

**Próximo paso:** `02_eda.ipynb` — Análisis Exploratorio de Datos